Revamping bookmarklet

Sequoia opened the pull request:
This change *inlines* ramda into the bookmarklet.  This is to say bookmarklet contains entire ramda script rather than trying to load remote `ramda.js`, which doesn't work in most places because of #1078 

I created a new `BOOKMARKLET.md`, which is built based on the latest `dist/ramda.js`. `npm run bookmarklet` does the following:

1. load `dist/ramda.js` into bookmarklet wrapper script (purpose of the latter being simply to show an "it worked" modal, possibly superfluous but I left it in)
2. load full bookmarklet script into `BOOKMARKLET.tmpl.md`
3. write that to `BOOKMARKLET.md`

I made put the bookmarklet its own MD file to avoid trashing up `README.md` history every time a build happens.

:point_right: [See it here](https://github.com/Sequoia/ramda/blob/bookmarklet/BOOKMARKLET.md) :point_left: 

# TODO
The build is a bit confusing, what with `make`, `npm` scripts, and `grunt`. I wasn't sure what your canonical "build" step is so I just added an npm script which can be included wherever.  `CONTRIBUTING.md` does not explain what the `build` step is either (unless "Run `make test lint`, (or `grunt test`) and address any errors." is the build step). **Please advise**, or just add the script in wherever your build is.

buzzdecafe commented:
hmmm, failing hint in the build. probably want to exclude that file from hint.

I think we would only want to update the README ad release-time, so we can add this script as a `prepublish` step. 

> The build is a bit confusing,

Totally agree, we're trying to standardize on just npm.

davidchambers commented:
First of all, this is very cool!

Secondly, the build process has always had quite a few moving parts, but @buzzdecafe has been working towards removing our dependencies on Make and Grunt (and its plethora of plug-ins).

Thirdly, there are a few stylistic issues you should address. To list them, run `npm run jshint` and `npm run jscs` after applying this patch:

```patch
diff --git a/package.json b/package.json
index a4f1aa0..3d111d5 100644
--- a/package.json
+++ b/package.json
@@ -36,8 +36,8 @@
     "browserify": "node_modules/browserify/bin/cmd.js test/*.js --outfile lib/test/bundle.js",
     "bookmarklet": "scripts/bookmarklet",
     "clean": "rimraf dist",
-    "jscs": "jscs scripts/build **/*.js",
-    "jshint": "jshint scripts/build **/*.js",
+    "jscs": "jscs scripts/{bookmarklet,build} **/*.js",
+    "jshint": "jshint scripts/{bookmarklet,build} **/*.js",
     "pretest": "npm run jshint && npm run jscs",
     "test": "node scripts/testfiles | node_modules/.bin/mocha --reporter spec"
   },
```

Finally, it would be great to incorporate this into our publishing process so the bookmarklet is automatically updated with each release. I can advise you as to how to go about this if you're interested. :)

Sequoia commented:
Thank you! In all my reading of the `CONTRIBUTING.md` I overlooked the lint/jscs stuff! :yum: Will correct now.

There *will* be a problem with the `bookmarklet.tmpl.js` because of the Handlebars token; I'll see if I can get around that with a jshint ignore comment.

Sequoia commented:
OK replaced last commit with new one. `jshint` & `jscs` tasks are passing. Also added `version` to bookmarklet "loaded" modal. I realized that there's otherwise no easy way to know if you've got a current version or not.

@davidchambers Yes I'm interested in adding this to the build, please do advise.

Incidentally, if you're planning to clean up the entire build that's something I might be able to help with... if so, open a ticket with a **list of goals/build tasks** (i.e. lint, minify, jscs, update docs, etc.). My approach would be to drop `make` & `grunt` and go [all npm run](http://substack.net/task_automation_with_npm_run) with build tasks written in JS where needed.

davidchambers commented on scripts/bookmarklet.tmpl.js:
```diff
@@ -0,0 +1,54 @@
+    ['text-align', 'center'],
+    ['border', 'solid 2px black'],
+    ['border-radius', '25px']
+  ]);
```
It seems to me this would be better as an object rather than an array of pairs. Also, why are some of the property names in camel case (e.g. `fontSize`) while others are in kebab case (e.g. `text-align`)?

Sequoia replied:
I made it an array just for simpler iteration. I'll switch it so it iterates over `Object.keys`. I'll make it all kebab as well.

davidchambers replied:
Since `setStyles` is only used in one place we could simply set these style properties directly:

```javascript
toast.style.position = 'fixed';
toast.style.top = '20px';
toast.style.right = '30%';
...
```

Also, why `right: 30%`? I'd have expected the notification to appear in the top-right corner.

davidchambers replied:
> I made it an array just for simpler iteration.

Don't forget `for ... in` loops. :)

```javascript
for (var prop in styles) {
  el.style[prop] = styles[prop];
}
```

As mentioned in my previous comment, though, I don't see what advantage `setStyles` provides over assigning the style properties directly in this case.

davidchambers commented on scripts/bookmarklet.tmpl.js:
```diff
@@ -0,0 +1,54 @@
+        jQuery(toast).fadeOut('slow', function() {
+          jQuery(this).remove();
+        });
+      }
```
I don't like branching based on whether jQuery is available. `b.removeChild(toast)` is fine. We needn't be fancy. :)

Sequoia replied:
Agreed. I just left that in from [before](https://github.com/ramda/ramda/commit/7b79d27cbc5aee5ad0c88cc91b70d268373ae579#diff-04c6e90faac2675aa89e2176d2eec7d8R91), will remove.

davidchambers commented on scripts/bookmarklet:
```diff
@@ -0,0 +1,26 @@
+var UglifyJS        = require('uglify-js');
+var fs              = require('fs');
+var path            = require('path');
+var pkg             = require(path.join(__dirname, '..', 'package.json'));
```
This could simply be `require('../package.json')`.

buzzdecafe commented:
> Incidentally, if you're planning to clean up the entire build that's something I might be able to help with... if so, open a ticket with a list of goals/build tasks (i.e. lint, minify, jscs, update docs, etc.). My approach would be to drop make & grunt and go all npm run with build tasks written in JS where needed.

On buzzdecafe/ramda where I am trying to do exactly what you describe: use only npm as our build tool. I would *gratefully* accept any help you have to offer. Part of this project also means moving docs stuff to ramda.github.io.

Sequoia commented:
1. changed to `require('../package.json');`
2. removed `setStyles` or whatever it was called
3. moved modal to top right
4. added bookmarklet build to `prepublish` script
5. sprinkled :sparkles: FP glitter :sparkles:  on `setTimeout` call
6. `jshint` & `jscs` pass

If there's more needed lemme know & I'll try to get to it.

buzzdecafe commented:
> sprinkled :sparkles: FP glitter :sparkles: on setTimeout call

that is my favorite part

Sequoia commented:
@davidchambers is your package.json patch relying on a type of shell expansion (not available on Travis)[https://travis-ci.org/ramda/ramda/jobs/62297836]?
```
> ramda@0.14.0 pretest /home/travis/build/ramda/ramda
> npm run jshint && npm run jscs


> ramda@0.14.0 jshint /home/travis/build/ramda/ramda
> jshint scripts/{bookmarklet,build} **/*.js

[31mERROR[0m: Can't open scripts/{bookmarklet,build}

> ramda@0.14.0 jscs /home/travis/build/ramda/ramda
> jscs scripts/{bookmarklet,build} **/*.js

Error: Path scripts/{bookmarklet,build} was not found.
    at /home/travis/build/ramda/ramda/node_modules/jscs/lib/checker.js:148:19
    at Array.0 (/home/travis/build/ramda/ramda/node_modules/jscs/node_modules/vow/lib/vow.js:699:56)
    at Object.callFns [as _onImmediate] (/home/travis/build/ramda/ramda/node_modules/jscs/node_modules/vow/lib/vow.js:23:35)
    at processImmediate [as _immediateCallback] (timers.js:354:15)
```

davidchambers commented on scripts/prepublish:
```diff
@@ -5,6 +5,7 @@ README="$(cat README.md)"
 README="${README//"$PREVIOUS_VERSION/ramda.min.js"/"$VERSION/ramda.min.js"}"
 README="${README//"${PREVIOUS_VERSION%.*}/ramda.min.js"/"${VERSION%.*}/ramda.min.js"}"
 echo "$README" >README.md
+npm run bookmarklet
```
We'll also need to add the changes after updating __BOOKMARKLET.md__:

```diff
 echo "$README" >README.md
 git add README.md

+npm run bookmarklet
+git add BOOKMARKLET.md
+
 rm -f dist/ramda{,.min}.js
 make dist/ramda{,.min}.js
 git add dist/ramda{,.min}.j
```

davidchambers commented:
Let's avoid the `{foo,bar}` expansion if it's causing problems. We can list each of the scripts individually.

Sequoia commented:
@davidchambers removed the `{a,b}` shell expansion, maybe CI will pass now :)

buzzdecafe commented:
:cow2:

CrossEye commented:
:herb:

davidchambers commented on scripts/prepublish:
```diff
@@ -5,7 +5,8 @@ README="$(cat README.md)"
 echo "$README" >README.md
-git add README.md
+npm run bookmarklet
+git add README.md BOOKMARKLET.md
```
Let's devote one paragraph to each file:

    paragraph for the creation
    and addition of README.md

    paragraph for the creation
    and addition of BOOKMARKLET.md

    paragraph for the creation
    and addition of dist/ramda{,.min}.js

In other words, leave the existing two paragraphs unchanged and add a new paragraph:

```bash
npm run bookmarklet
git add BOOKMARKLET.md
```

I'm sorry that I didn't make this clear in my previous comment.

Sequoia replied:
np

Sequoia commented:
@davidchambers bookmarklet build needed to come AFTER `dist/` update. done now.

davidchambers commented:
> bookmarklet build needed to come AFTER `dist/` update

Good point! Thanks for the update. :)
